In [1]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [3]:
#change the file location according to your file location in google drive
import os
os.chdir('/content/drive/My Drive/Final_Project_CS498/Filtered_Data_from_Illinois/')

In [4]:
# first neural network with keras tutorial
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from keras.utils import np_utils
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import LabelEncoder
from sklearn.pipeline import Pipeline


In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [6]:
train_df= pd.read_csv("train_data_normalizedv3.csv")

In [7]:
test_df=pd.read_csv("Test_data_normalized_v3.csv")

In [ ]:
#train_df=train_df.iloc[1001:,:]

In [ ]:
#test_df=test_df.append(train_df.iloc[0:1000,:], ignore_index = True)

In [ ]:
#test_df.drop(columns='Unnamed: 0', inplace= True)

In [ ]:
#train_df.drop(columns='Unnamed: 0', inplace= True)

In [ ]:
#test_df.to_csv("test_normalizedv2.csv", index=0)

In [ ]:
#train_df.to_csv("train_normalizedv2.csv", index=0)

In [8]:
train_df.head()

Unnamed: 0      aadt  access  ...  spd_limt  surf_cat  Accident
0           0 -0.479109       0  ...  0.043635         1       0.0
1           1 -0.510306       0  ...  0.043635         1       0.0
2           2 -0.110980       0  ... -0.274040         1       0.0
3           3 -0.524865       0  ...  0.043635         1       0.0
4           4 -0.418794       0  ...  0.043635         1       0.0

[5 rows x 14 columns]

In [9]:
X=train_df.iloc[:,1:-1]
y=train_df.iloc[:, -1]

In [10]:
X.head()

aadt  access  curv_rad_binary  ...   seg_lng  spd_limt  surf_cat
0 -0.479109       0                0  ...  0.043635  0.043635         1
1 -0.510306       0                0  ...  0.043635  0.043635         1
2 -0.110980       0                0  ... -0.274040 -0.274040         1
3 -0.524865       0                0  ...  0.043635  0.043635         1
4 -0.418794       0                0  ...  0.043635  0.043635         1

[5 rows x 12 columns]

In [11]:
len(X.columns)

12

In [12]:
dummy_y = np_utils.to_categorical(y)

In [14]:
X_test=test_df.iloc[:,1:-1]
y_test=test_df.iloc[:,-1]

In [16]:
dummy_y_test=np_utils.to_categorical(y_test)

model

In [ ]:
epochs =[20,30]
batch_sizes=[50,100]

In [17]:

# define baseline model
def baseline_model():
	model = Sequential()
	model.add(Dense(8, input_dim=12, activation='relu'))
	model.add(Dense(4, activation='softmax'))
	model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
	return model

In [21]:
def baseline_model1():
  model=Sequential()
  model.add(Dense(12, input_dim=12, activation='relu'))
  model.add(Dense(8, activation='relu'))
  model.add(Dense(4, activation='softmax'))
  model.compile(loss= 'categorical_crossentropy', optimizer= 'adam', metrics =['accuracy'])
  return model

In [ ]:
model = Sequential()
model.add(Dense(50, input_dim=12, activation='relu'))
model.add(Dense(4, activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(X, dummy_y, epochs=20, batch_size=50, verbose=0)
# evaluate the model
scores = model.evaluate(X_test, dummy_y_test, verbose=0)
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))
 
# serialize model to JSON
#model_json = model.to_json()
#with open("model.json", "w") as json_file:
#    json_file.write(model_json)
# serialize weights to HDF5
#model.save_weights("model.h5")
#print("Saved model to disk")'''

accuracy: 70.80%


**Without k-fold**

1 layer:

8 nodes: 70.60%

12 nodes: 69.90%

24 nodes: 71.05%

30 nodes: 71.15%

50 nodes: 70.80%



In [22]:
estimator = KerasClassifier(build_fn=baseline_model1, epochs=20, batch_size=50, verbose=0)
kfold = KFold(n_splits=5, shuffle=True)
results = cross_val_score(estimator, X, dummy_y, cv=kfold)
print("Baseline: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))

Baseline: 73.62% (0.13%)


In [ ]:
result1=estimator.evaluate(X_test, y_test)
print(result1)

AttributeError: ignored

kfold

1 layer 8 nodes: 73.59%

1 layer 12 nodes: 73.60%

2 layer 8,10 nodes: 73.62% 

2 layer 12, 8 nodes: 73.62%

#old model
**1 hidden layer of 8 nodes**

epoch =20, batch = 50, accuracy: 73.26% (0.15%)

epoch = 20, batch = 100, accuracy: 73.22% (0.12%)

epoch = 30, batch = 50, accuracy: 73.23% (0.12%)

epoch = 30, batch = 100, accuracy: 73.21% (0.13%)



**1 hidden layer of 12 nodes**

epoch =20, batch = 50, accuracy:  73.20% (0.14%)


**1 hidden layer of 16 nodes**

epoch = 20, batch = 50, accuracy: 73.29% (0.11%)


**1 hidden layer of 24 nodes**

epoch = 20, batch = 50, accuracy: 73.28% (0.11%)




**2 hidden layers( 8,4)**

epoch = 20, batch = 100, accuracy: 73.23% (0.05%)

epoch =50, batch =100, accuracy: 73.28% (0.13%)

epoch = 100, batch =100, accuracy: 73.26% (0.10%)